Print help even if logged out

chemotaxis opened the pull request:
Fixes #3823.

**Testing note**:  I logged out first using `gh auth logout`.

Currently, if you're logged out, you can't access any help pages with the flags `--help` and `-h` for commands that require authentication.

While `gh help actions` works, `gh actions --help` does not.

If you try `gh actions --help`, you get a message about authentication:

```text
Welcome to GitHub CLI!

To authenticate, please run `gh auth login`.
```

This pull request allows you to explicitly ask for help pages using the help flags, even if you are logged out.

After applying this change, `gh actions` will still print the same message above, but `gh actions --help` will now print this:

```text
Welcome to GitHub Actions on the command line.

GitHub CLI integrates with Actions to help you manage runs and workflows.

Interacting with workflow runs  
gh run list:      List recent workflow runs  
gh run view:      View details for a workflow run or one of its jobs  
gh run watch:     Watch a workflow run while it executes  
gh run rerun:     Rerun a failed workflow run  
gh run download:  Download artifacts generated by runs

[Rest of help page]...
```

chemotaxis committed:
Ask for and print help even if logged out

You have to explicitly ask for help using the help flags.  Otherwise,
`gh` will just print the authentication message.

mislav reviewed (changes requested):
Could this entire check be moved into `cmdutil.IsAuthCheckEnabled` using Cobra introspection features?

1. If the command to be executed has the name `help`, auth check should be skipped. This would handle `gh help <command>`
2. If the command to be executed has a `--help` flag (as determined by `cmd.Flags().Lookup("help")`) and the flag is enabled, auth check should be skipped. This will transparently handle `-h` as well.

chemotaxis commented:
> 1. If the command to be executed has the name `help`, auth check should be skipped. This would handle `gh help <command>`

`gh help actions` already skips authentication and returns the help page, even if you're logged out.  `gh` returns `false` for `cmdutil.IsAuthCheckEnabled()`.

> 2. If the command to be executed has a `--help` flag (as determined by `cmd.Flags().Lookup("help")`) and the flag is enabled, auth check should be skipped. This will transparently handle `-h` as well.

When I run `gh actions --help` with this code:

```go
func IsAuthCheckEnabled(cmd *cobra.Command) bool {
	log.SetFlags(log.Llongfile)
	log.Printf("%+v", cmd.Flags().Lookup("help"))
	log.Printf("%+v", cmd.Flags())
...
```

I get:

```
github.com/cli/cli/pkg/cmdutil/auth_check.go:43: <nil>
github.com/cli/cli/pkg/cmdutil/auth_check.go:44: &{
  Usage:<nil> 
  SortFlags:true 
  ParseErrorsWhitelist:{UnknownFlags:false} 
  name:actions 
  parsed:false 
  actual:map[] 
  orderedActual:[] 
  sortedActual:[] 
  formal:map[] 
  orderedFormal:[] 
  sortedFormal:[] 
  shorthands:map[] 
  args:[] 
  argsLenAtDash:-1 
  errorHandling:0 
  output:0xc0003a7e30 
  interspersed:true 
  normalizeNameFunc:<nil> 
  addedGoFlagSets:[]
}
```

I tried looking for an appropriate Cobra method, but it seems the flags aren't getting set somewhere.  Am I missing something obvious?  If I wasn't depending on `expandedArgs` in `gh/main.go`, I could move the check into `cmdutil.IsAuthCheckEnabled`.

chemotaxis committed:
Skip authentication message if asking for help

Currently, this still checks authentication, but we skip the
authentication message and exit normally.

chemotaxis committed:
Add comment about cmd.Help()

The linter picked up that the error value from cmd.Help() isn't checked.

Even though cmd.Help() returns an error value, it's always nil. The
inner HelpFunc() function directly prints the error message instead of
returning an error value.

Strawberry14388 reviewed (approved):
Add

chemotaxis commented:
I've updated the pull request and the pull request description.  My latest commits now skip the authentication message and just print the help page.  We still do the authentication check though.

I've dug around a little more, but I'm still not sure why `cmd.Flags().Lookup("help")` returns `nil`.

mislav committed:
Add a global pre-run hook to handle auth check and repo override

With auth check being done via Cobra hooks, it is automatically skipped
for non-runnable commands and `-h/--help` flag usage.

mislav committed:
Simplify `gh actions` implemenation

The command is now non-runnable, meaning it's exempt from auth check.

mislav committed:
Merge remote-tracking branch 'origin' into fix-actions-help

mislav committed:
Fix repo override

mislav commented:
Thanks for looking into all this. I didn't realize Cobra didn't make this more straightforward. I've pushed some changes to avoid manually checking for `-h/--help` flags and to avoid re-implementing the help system. It relies on Cobra's pre-run hooks. Since Cobra's implementation of hooks is pretty bad, I had to jump through hoops to ensure that the `-R/--repo` override hook doesn't accidentally override the auth-checking one.

chemotaxis committed:
Whitespace

chemotaxis commented:
No problem!  I'm still trying to understand Cobra and the codebase, so I wasn't sure if I was missing something obvious.

The latest commits add back the help "footer" that shows usage, inherited flags, etc.  At the moment, it doesn't really matter if the footer is there or not, but if you add other flags to `gh actions` later, it should appear.  Plus, the footer appears in the `trunk` branch.

chemotaxis committed:
Restore help footer

At the moment, the "help footer" doesn't add any new information, but if
additional flags are added later, they should appear in the footer. This
change restores this help footer:

```shell
USAGE
  gh actions [flags]

INHERITED FLAGS
  --help   Show help for command

LEARN MORE
  Use 'gh <command> <subcommand> --help' for more information about a command.
  Read the manual at https://cli.github.com/manual
```

chemotaxis commented:
No problem! I'm still trying to understand Cobra and the `gh` codebase, so I wasn't sure if I was missing something obvious.

The latest commit adds back the help "footer" that shows usage, inherited flags, etc. At the moment, the footer doesn't provide any new information, but if you add other flags to `gh actions` later, it should appear.

mislav reviewed (approved):
Thank you for your help!

mislav committed:
Revert "Whitespace"

Trailing whitespace is significant in Markdown.

This reverts commit 682c15d52c387e4af9d065ce8c536a45f813a6a7.

mislav commented:
@chemotaxis I've reverted whitespace changes to the help doc because we feed all docs as Markdown to our documentation site, and whitespace (even trailing) in Markdown is significant.

chemotaxis commented:
> @chemotaxis I've reverted whitespace changes to the help doc because we feed all docs as Markdown to our documentation site, and whitespace (even trailing) in Markdown is significant.

Sorry, I didn't realize that!

vilmibm commented on pkg/cmd/actions/actions.go:
```diff
@@ -1,30 +1,19 @@
-		Run: func(cmd *cobra.Command, args []string) {
-			actionsRun(opts)
-		},
+		Long:  actionsExplainer(cs),
```
If we're not going to be in the business of passing `nil` to `actionsExplainer` anymore, please clean up `actionsExplainer` to not handle that case.

I feel like I set it up that way to avoid some issue with rendering the docs but I just tested rendering them locally with both this branch and `trunk` and it seems ok; still, I'll make a point of checking that once this is released.

hamad22h replied:
ا

chemotaxis replied:
I removed the `nil` check.

chemotaxis committed:
Always try to render bold font

It looks like a similar check is done in ColorScheme.Bold() where it
checks whether the scheme is enabled or not.

chemotaxis committed:
Disable authentication check, but keep runnable

In this branch, we originally avoided the authentication check by
getting rid of the run method attached to the command.  Instead of that,
this commit makes the `gh actions` command runnable again, but the
authentication is disabled with `cmdutil.DisableAuthCheck`; this mirrors
what's done for `gh version`.

`gh actions` and `gh actions [-h | --help]` all work while being logged
out.

In addition, this commit restores some original behavior.  Before this
commit, the help footer (usage, inherited flags, etc.) is appended
whether you use `gh actions` or `gh actions --help`.  This commit
restores the original behavior where `gh actions` prints just the text
for the actions explanation, but `gh actions --help` appends the help
footer.

chemotaxis commented:
This pull request now has two main fixes: 

- Fixing the general problem of bypassing authentication when explicitly asking for help using any command
  - Caveat: `gh help <command>` actually already works
- Refactoring `gh actions`

I was looking over the changes one last time and noticed the `cmdutil.DisableAuthCheck` function.  I thought I'd gone over it before, but decided to investigate uses of it in the codebase.  `gh actions` can bypass authentication by just adding `cmdutil.DisableAuthCheck` to the command instead of making it non-runnable.

The latest commit makes `gh actions` runnable again, but I've disabled authentication with the disable function.  There are quite a few uses of the disable function, but `gh version` is probably the simplest example.

This also restores the original print behavior where `gh actions` prints just the text for explaining actions and workflows, but `gh actions --help` appends the help footer (usage, inherited flags, etc.).

Feel free to revert b0f58d0, but I think this is a better solution since it maintains the original print behavior for `gh actions`.

samcoe reviewed (approved):
This looks good to me

vilmibm reviewed (approved):
thank you!
